README.md
# Connect DynamoDB

[![CI](https://github.com/samuraitruong/connect-dynamodb/actions/workflows/release.yml/badge.svg)](https://github.com/samuraitruong/connect-dynamodb/actions/workflows/release.yml)

This project was forked from https://github.com/ca98am79/connect-dynamodb and upgrade the code to moderm syntax + devtools. Going forward, I wil maintain and keep library up to date

connect-dynamodb is a DynamoDB session store backed by the [aws-sdk](https://github.com/aws/aws-sdk-js)

** This version is modified to allow read database session value and merge with new change to avoid race condition overriten session data **

## Installation

```sh
    npm install @samuraitruong/connect-dynamodb
    or
    yarn add @samuraitruong/connect-dynamodb
```

## Options

Rational defaults are set but can be overridden in the options object. Credentials and configuration are automatically loaded from [environment variables](http://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/loading-node-credentials-environment.html) or [shared credentials](http://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/loading-node-credentials-shared.html) but may optionally be passed through a JSON file or object. The client attribute is necessary for use with [DynamoDB Local](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html) but can be left out if using DynamoDB with your AWS account. To use [DynamoDB TTL](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/TTL.html), enable it on the table and select the `expires` field.

- One of the following if not using environment variables or shared credentials:
  - `AWSConfigPath` Optional path to a [file containing your AWS credentials and configuration](http://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html#Credentials_from_Disk)
  - `AWSConfigJSON` Optional [JSON object containing your AWS credentials and configuration](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Config.html)
- `client` Optional AWS DynamoDB object from `new AWS.DynamoDB()`
- `AWSRegion` Optional AWS region (defaults to 'us-east-1', ignored if using `AWSConfigPath` or `AWSConfigJSON`)
- `table` Optional DynamoDB server session table name (defaults to "sessions")
- `hashKey` Optional hash key (defaults to "id")
- `prefix` Optional key prefix (defaults to "sess")
- `reapInterval` Legacy session expiration cleanup in milliseconds. Should instead enable [DynamoDB TTL](http://docs.aws.amazon.com/amazondynamodb/latest/developerguide/TTL.html) and select the `expires` field.
- `ttl`: the values in second that will set for `ttlFieldName` field default = 365 days
- `ttlFieldName`: set the column name for ttl, default = ttl
- **BREAKING CHANGE** from v1.0.11 to v2.0.0 for reaping sessions with changes to the format of the expires field timestamp.

## Usage

````js
    const options = {
        // Optional DynamoDB table name, defaults to 'sessions'
        table: 'myapp-sessions',

        // Optional path to AWS credentials and configuration file
        // AWSConfigPath: './path/to/credentials.json',

        // Optional JSON object of AWS credentials and configuration
        AWSConfigJSON: {
            accessKeyId: <YOUR_ACCESS_KEY_ID>,
            secretAccessKey: <YOUR_SECRET_ACCESS_KEY>,
            region: 'us-east-1'
        },

        // Optional client for alternate endpoint, such as DynamoDB Local
        client: new AWS.DynamoDB({ endpoint: new AWS.Endpoint('http://localhost:8000')}),

        // Optional ProvisionedThroughput params, defaults to 5
        readCapacityUnits: 25,
        writeCapacityUnits: 25
        // TTL default set to 1 year, below to set ttl to 30 days
        ttl: 30 * 3600 * 24
    };
    ```

### with connect  [connect](https://github.com/senchalabs/connect)
```js
    var connect = require('connect');
    var DynamoDBStore = require('connect-dynamodb')(connect);
    connect()
        .use(connect.cookieParser())
        .use(connect.session({ store: new DynamoDBStore(options), secret: 'keyboard cat'}));
````

### With [express 3](http://expressjs.com/en/3x/api.html)

```js
    var DynamoDBStore = require('connect-dynamodb')(express);
    var app = express(
        express.cookieParser(),
        express.session({ store: new DynamoDBStore(options), secret: 'keyboard cat'});
    );
```

### With [express 4](http://expressjs.com/)

```js
var app = express();
var session = require('express-session');
var DynamoDBStore = require('connect-dynamodb')({ session: session });
app.use(session({ store: new DynamoDBStore(options), secret: 'keyboard cat' }));

OR;

var app = express();
var session = require('express-session');
var DynamoDBStore = require('connect-dynamodb')(session);
app.use(session({ store: new DynamoDBStore(options), secret: 'keyboard cat' }));
```

## Testing

If you want to run the tests locally and have the AWS environment variables setup you can run the command:

```
npm test
```

You can also run it locally by running the following two scripts in separate terminals:

```
docker run -it --rm -d \
  --name=dynamodb-test \
  -p 127.0.0.1:8000:8000 \
  amazon/dynamodb-local
```

```
export AWS_CONFIG_JSON='{"endpoint": "http://127.0.0.1:8000", "region": "us-east-1", "accessKeyId": "accesskey", "secretAccessKey": "secretaccesskey"}'
npm test
```

## IAM Permissions

Connect DynamoDB requires the following IAM permissions for DynamoDB:

- CreateTable
- PutItem
- DeleteItem
- GetItem
- Scan
- UpdateItem

Sample IAM policy (with least privilege):

_(Replace **\<AWS ACCOUNT ID\>**, **\<TABLE NAME\>** and **\<SOURCE IP AND BITMASK\>**)._

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "dynamodb:CreateTable",
                "dynamodb:PutItem",
                "dynamodb:DeleteItem",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:UpdateItem"
            ],
            "Resource": "arn:aws:dynamodb:*:<AWS ACCOUNT ID>:table/<TABLE NAME>"
        }
    ]
}
```

## License

connect-dynamodb is licensed under the [MIT license.](https://github.com/ca98am79/connect-dynamodb/blob/master/LICENSE.txt)

## Credits

All credeit to creator and maintainer of original repo (https://github.com/ca98am79/connect-dynamodb)[https://github.com/ca98am79/connect-dynamodb]


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.2.0](https://github.com/samuraitruong/connect-dynamodb/compare/v2.1.7...v2.2.0) (2021-09-07)


### Features

* add support separate field for ttl ([849ccae](https://github.com/samuraitruong/connect-dynamodb/commit/849ccaedad7c9e208c517cbcbdec24bbab69d725))

### [2.1.7](https://github.com/samuraitruong/connect-dynamodb/compare/v2.1.6...v2.1.7) (2021-07-26)


### Bug Fixes

* return value and update touch test for backwards compatibility ([ace121a](https://github.com/samuraitruong/connect-dynamodb/commit/ace121ae5fc86229d2918d8c5d2a1433caa8e6c8))
* touch implementation should use callback ([f485ee9](https://github.com/samuraitruong/connect-dynamodb/commit/f485ee93f84650fdbde15784c7a773a191de2baa))

### [2.1.6](https://github.com/samuraitruong/connect-dynamodb/compare/v2.1.5...v2.1.6) (2021-07-26)

### [2.1.5](https://github.com/samuraitruong/connect-dynamodb/compare/v2.1.4...v2.1.5) (2021-07-23)


### Bug Fixes

* remove lastRead which cause session always change ([87fba99](https://github.com/samuraitruong/connect-dynamodb/commit/87fba99a299e05e0787d58a96cf0d3bedce030ca))

### [2.1.4](https://github.com/samuraitruong/connect-dynamodb/compare/v2.1.3...v2.1.4) (2021-07-12)


### Bug Fixes

* initialize create table on construct ([790b5d2](https://github.com/samuraitruong/connect-dynamodb/commit/790b5d2741767d7d4efdffb8316ae5005162f9c3))
* table existing ([6915a7e](https://github.com/samuraitruong/connect-dynamodb/commit/6915a7e420e0cee25f6c985f67fde4b960616167))

### [2.1.3](https://github.com/samuraitruong/connect-dynamodb/compare/v2.1.2...v2.1.3) (2021-07-01)

### [2.1.2](https://github.com/samuraitruong/connect-dynamodb/compare/v2.1.1...v2.1.2) (2021-07-01)

### [2.1.1](https://github.com/samuraitruong/connect-dynamodb/compare/v2.1.0...v2.1.1) (2021-06-30)

## [2.1.0](https://github.com/samuraitruong/connect-dynamodb/compare/v1.0.4...v2.1.0) (2021-06-30)


### Features

* refactoring the code to typescipt and module ([f405851](https://github.com/samuraitruong/connect-dynamodb/commit/f405851a064ba756173fa29cb7a533045ef15520))


### Bug Fixes

* correct check server side expires session ([5a1aba2](https://github.com/samuraitruong/connect-dynamodb/commit/5a1aba2de2d8ef3305bc9dc49742970456d2935a))

2.0.4 / 2019-11-15

  * Don't crash on invalid session data stored in DDB (https://github.com/ca98am79/connect-dynamodb/issues/44)

2.0.3 / 2019-01-02

  * Support passing express-session directly to the adapter (https://github.com/ca98am79/connect-dynamodb/issues/54)

2.0.2 / 2018-09-12

  * fix to use the hashKey provided in config

2.0.1 / 2018-07-07

  * Make possible to use npm install --no-optional

2.0.0 / 2017-08-19
==================

  * 1.0.12 had breaking change with dates

1.0.12 / 2017-07-03
==================

  * Implemented touch as requested on #23 and changed the expires field to work on seconds units instead of millis to address #39. This way to use TTL we just need to enable it on the table and select the 'expires' field. (https://github.com/ca98am79/connect-dynamodb/pull/43)

1.0.11 / 2017-03-14
==================

  * Configurable ProvisionedThroughput read/write capacity units (https://github.com/ca98am79/connect-dynamodb/pull/40)

1.0.10 / 2017-02-27
==================

  * Set Consistent read by default (https://github.com/ca98am79/connect-dynamodb/pull/37)

1.0.9 / 2016-08-21
==================

  * Fix bug with reap method (https://github.com/ca98am79/connect-dynamodb/issues/33)

1.0.8 / 2016-07-18
==================

  * Fix issue with reap method (https://github.com/ca98am79/connect-dynamodb/issues/27)

1.0.7 / 2016-01-29
==================

  * Add ability to provide a JSON object to configure AWS, specify custom hashKey (https://github.com/ca98am79/connect-dynamodb/pull/26)

1.0.6 / 2014-01-29
==================

  * bug fix (https://github.com/ca98am79/connect-dynamodb/issues/16)

1.0.5 / 2014-01-28
==================

  * Change to allow loading of AWS credentials from environment vars (https://github.com/ca98am79/connect-dynamodb/issues/15)

1.0.4 / 2013-07-27
==================

  * Fix for default table name (https://github.com/ca98am79/connect-dynamodb/issues/13)

1.0.3 / 2013-08-10
==================

  * Change tests to should.js

1.0.2 / 2013-07-27
==================

  * Properly handle errors in get (https://github.com/ca98am79/connect-dynamodb/issues/12)

1.0.1 / 2013-06-27
==================

  * Change expires check to a number (https://github.com/ca98am79/connect-dynamodb/issues/11)

1.0.0 / 2013-05-25
==================

  * Adding support for the 2012-08-10 API version of DynamoDB (https://github.com/ca98am79/connect-dynamodb/issues/6)

0.1.4 / 2013-04-19
==================

  * Specify API version because of breaking updates to aws-sdk-js (http://aws.amazon.com/releasenotes/JavaScript/3118686131819314)

0.1.3 / 2013-04-11
==================

  * Fix expiry default if no sess.cookie.maxAge (https://github.com/ca98am79/connect-dynamodb/issues/4)

0.1.2 / 2013-04-06
==================

  * Fixed reap issue (https://github.com/ca98am79/connect-dynamodb/issues/3)

0.1.1 / 2013-01-26
==================

  * Cleanup of tests 

0.1.0 / 2013-01-25
==================

  * Switch to use aws-sdk instead of dynode
  
0.0.6 / 2012-06-29
==================

  * Fix an error with dynode 0.5.0
  
0.0.5 / 2012-06-13
==================

  * Added session table check/creation at init

0.0.2 / 2012-05-23
==================

  * Added reap deletion of expired sessions
  
0.0.1 / 2012-05-21
==================

  * Initial commit


src/__test__/store.test.ts
import AWS from 'aws-sdk';
import session from 'express-session';
import connect from '../index';
const DynamoDBStore = connect({ session: session });

const client = new AWS.DynamoDB({ endpoint: "http://localhost:8000", region: 'us-east-1' });

const store = new DynamoDBStore({
    client,
    table: 'sessions-test',
    ttl: 5000
});

describe('DynamoDBStore', function () {

    describe('Setting', function () {
        it('should store data correctly', async () => {

            await store.set('123', {
                cookie: {
                    maxAge: 2000
                },
                name: 'tj'
            }, (err) => {
                expect(err).toEqual(null)
            });
        });

        it('should set ttl value ', async () => {
            const now = Math.floor(new Date().getTime() / 1000);

            await store.set('789', {
                cookie: {
                    maxAge: 2000
                },
                name: 'test'
            }, (err) => {
                expect(err).toEqual(null)
            });
            const sess = await store.get('789', (err, data) => data);
            expect(sess.__ttl__).toBeGreaterThanOrEqual(now);
            expect(sess.__ttl__).toBeLessThanOrEqual(now + 5000)
        });


    });
    describe('Getting', function () {

        beforeAll(async () => {
            await store.set('1234', {
                cookie: {
                    maxAge: 2000
                },
                name: 'tj'
            }, jest.fn());
        });


        it('should get data correctly', async () => {

            await store.get('1234', (err, res) => {
                if (err) throw err;
                expect(res.cookie).toEqual({
                    maxAge: 2000
                });
                expect(res.name).toEqual('tj');

            });
        });

        it('does not crash on invalid session object', async () => {

            await store.get('error-session', (err, res) => {
                if (err) throw err;
                expect(res).toEqual(undefined)
            });
        });

    });
    describe('Touching', function () {
        const sess = {
            cookie: {
                maxAge: 2000
            },
            name: 'tj'
        };
        let maxAge = 0;
        beforeEach(async () => {

            maxAge = (Math.floor((Date.now() + 2000) / 1000));
            await store.set('1234', sess, jest.fn());
        });

        it('should touch data correctly', (done) => {
            setTimeout(() => {
                // express-session uses callback instead of promise
                // @see https://github.com/expressjs/session#storetouchsid-session-callback
                store.touch('1234', sess, (err, res) => {
                    expect(err).toBeNull();
                    const expires = +res.Attributes.expires.N;
                    expect(expires).toBeGreaterThan(maxAge);
                    done();
                });
            }, 2000);
        });

        it('should also support promise for backwards compatibility', async () => {
            await new Promise(r => setTimeout(r, 2000));
            const res: any = await store.touch('1234', sess);
            const expires = +res.Attributes.expires.N;
            expect(expires).toBeGreaterThan(maxAge);
        });

    });

    describe('Destroying', function () {
        beforeAll(async () => {
            await store.set('12345', {
                cookie: {
                    maxAge: 2000
                },
                name: 'tj'
            }, jest.fn());
        });

        it('should destroy data correctly', async () => {
            const store = new DynamoDBStore({
                client,
                table: 'sessions-test'
            });
            const test = await store.get('12345', (err, res) => res);
            expect(test).not.toEqual(null);

            await store.destroy('12345');

            const again = await store.get('12345', (err, res) => res);
            expect(again).toBe(undefined);
        });

    });
    // describe('Reaping', function () {
    //     before(async() => {
    //         const store = new DynamoDBStore({
    //             client,
    //             table: 'sessions-test'
    //         });
    //         store.set('123456', {
    //             cookie: {
    //                 maxAge: -20000
    //             },
    //             name: 'tj'
    //         }, done);
    //     });

    //     it('should reap data correctly', async() => {
    //         this.timeout(5000); // increased timeout for local dynamo
    //         const store = new DynamoDBStore({
    //             client,
    //             table: 'sessions-test'
    //         });
    //         store.reap(function (err, res) {
    //             if (err) throw err;

    //             store.get('123456', function (err, res) {
    //                 if (err) throw err;
    //                 should.not.exist(res);

    //                 done();
    //             });
    //         });
    //     });

    // });
});


src/__test__/express.test.ts
import request from 'supertest';
import { app } from './app';
jest.setTimeout(30000)
describe('Test connect', () => {

  it('should successful request', (done) => {
    request(app())
      .get('/')
      .set('Accept', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200, done);
  });

  it('should not keep session for multiple call with different session', (done) => {
    const server = app();
    let cookies: any;
    const agent = request(server);

    agent.get('/inc')
      .expect((res) => {
        cookies = res.headers['set-cookie'];
        expect(res.body.count).toEqual(1)
      })
      .end(() => {
        agent.get('/inc')
          .expect((res) => {
            expect(res.body.count).toEqual(1)
          })
          .expect(200, done)
      })
  });

  it('should maintain session for multiple call with same session', (done) => {
    const server = app();
    let cookies: any;
    const agent = request(server);

    agent.get('/inc')
      .expect((res) => {
        cookies = res.headers['set-cookie'];
        expect(res.body.count).toEqual(1)
      })
      .end(() => {
        agent.get('/inc')
          .set('Cookie', cookies)
          .expect((res) => {
            expect(res.body.count).toEqual(2)
          })
          .expect(200, done)
      })
  });

  it('should reset session when generate was called', (done) => {
    const server = request(app());
    let cookies: any;
    server
      .get('/inc')
      .expect((res) => {
        cookies = res.headers['set-cookie']
        expect(res.body.count).toEqual(1)
      })
      .end(() => {
        server
          .get('/reset')
          .set('Cookie', cookies)
          .expect((res) => {
            cookies = res.headers['set-cookie']
          })
          .end(() => {
            server
              .get('/inc')
              .set('Cookie', cookies)
              .expect((res) => {
                cookies = res.headers['set-cookie']
                expect(res.body.count).toEqual(1)
              })
              .end(done)
          })
      })



  });


  it('should not return expired session', (done) => {
    const server = request(app(2000));
    let cookies: any;
    server
      .get('/inc')
      .expect((res) => {
        cookies = res.headers['set-cookie']
        expect(res.body.count).toEqual(1)
      })
      .end(() => {
        setTimeout(() => {
          server
            .get('/inc')
            .set('Cookie', cookies)
            .expect((res) => {
              cookies = res.headers['set-cookie']
              expect(res.body.count).toEqual(1)
            })
            .end(done)
        }, 2001)
      })
  });

})


src/store.ts
/*!
 * Connect - DynamoDB
 * Copyright(c) 2020 introvert.com LLC <support@introvert.com>, forked and upgrade by samuraitruong
 * MIT Licensed
 */

import AWS from 'aws-sdk';
import session from 'express-session';
import { oneDayInMilliseconds } from './constants';

export class DynamoDBStore extends session.Store {
  private connect: any = {};
  private options: any;

  private prefix: string;
  private hashKey: string;
  private readCapacityUnits: number;
  private writeCapacityUnits: number;
  private client: AWS.DynamoDB;
  private table: string;
  private reapInterval: number;
  private _reap: any;
  private _tableInfo: any = null;

  /**
  * Initialize DynamoDBStore with the given `options`.
  *
  * @param {Object} options
  * @api public
  */
  constructor(options: any = {}) {
    super(options);

    this.options = options || {};
    this.options.ttl = this.options.ttl || 365 * 24 * 3600; // defaut to 1 years
    this.options.ttlFieldName = this.options.ttlFieldName || 'ttl'; // defaut to 1 years
    this.prefix = null == options.prefix ? 'sess:' : options.prefix;
    this.hashKey = null == options.hashKey ? 'id' : options.hashKey;
    this.readCapacityUnits = parseInt(options.readCapacityUnits || 5, 10);
    this.writeCapacityUnits = parseInt(options.writeCapacityUnits || 5, 10);

    if (options.client) {
      this.client = options.client;
    } else {
      if (options.AWSConfigPath) {
        AWS.config.loadFromPath(options.AWSConfigPath);
      } else if (options.AWSConfigJSON) {
        AWS.config.update(options.AWSConfigJSON);
      } else {
        const region = options.AWSRegion || 'us-east-1';
        AWS.config.update({ region });
      }
      this.client = new AWS.DynamoDB();
    }

    this.table = options.table || 'sessions';
    this.reapInterval = options.reapInterval || 0;
    if (this.reapInterval > 0) {
      this._reap = setInterval(this.reap.bind(this), this.reapInterval);
    }
    this.prepareTable = this.prepareTable.bind(this);
    this.prepareTable();
  }

  async prepareTable() {
    if (this._tableInfo) return;
    try {
      const info = await this.client.describeTable(
        {
          TableName: this.table,
        }).promise();
      this._tableInfo = info;
      // console.log(JSON.stringify(info, null, 4))
    } catch (err) {
      try {
        this._tableInfo = await this.client.createTable(
          {
            TableName: this.table,
            AttributeDefinitions: [
              {
                AttributeName: this.hashKey,
                AttributeType: 'S',
              }
            ],
            KeySchema: [
              {
                AttributeName: this.hashKey,
                KeyType: 'HASH',
              },
            ],
            ProvisionedThroughput: {
              ReadCapacityUnits: this.readCapacityUnits,
              WriteCapacityUnits: this.writeCapacityUnits,
            },
          }).promise();
      }
      catch (err) {
        if (err.code !== 'ResourceInUseException') {
          console.log(err);
          throw err;
        }
      }
    }

  }

  /**
   * Attempt to fetch session by the given `sid`.
   *
   * @param {String} sid
   * @param {Function} fn
   * @api public
   */
  async get(sid: string, fn: (err?: any, data?: any) => any) {
    await this.prepareTable();
    const now = Math.floor(Date.now() / 1000);
    const params = {
      TableName: this.table,
      Key: {
        [this.hashKey]: {
          S: this.prefix + sid,
        }
      },
      ConsistentRead: true,
    };
    try {
      const result = await this.client.getItem(params).promise();
      if (!(result.Item && result.Item.sess && result.Item.sess.S))
        return fn(null);
      else if (result.Item.expires && now >= +(result.Item.expires?.N || 0)) {
        return fn(null);
      } else {
        const sessStr = result.Item.sess.S.toString();
        const sess = JSON.parse(sessStr);
        if (result.Item && result.Item.ttl) {
          sess.__ttl__ = +(result.Item[this.options.ttlFieldName].N as string)
        }
        return fn(null, sess);
      }
    }
    catch (err) {
      return fn(err)
    }
  }

  /**
   * Commit the given `sess` object associated with the given `sid`.
   *
   * @param {String} sid
   * @param {Session} sess
   * @param {Function} fn
   * @api public
   */
  async set(sid: string, currentSess: any, fn: (err?: any) => any) {
    await this.prepareTable();
    try {
      const oldSession = await this.get(sid, (err, data) => data);
      const newSess = Object.assign({}, oldSession, currentSess);

      const data = await this.setInternal(sid, newSess);
      fn(null)
    } catch (err) {
      fn(err)
    }
  }


  async setInternal(sid: string, sess: any) {
    const expires = this.getExpiresValue(sess);
    sess = JSON.stringify(sess);
    const params = {
      TableName: this.table,
      Item: {
        [this.hashKey]: {
          S: this.prefix + sid,
        },
        expires: {
          N: JSON.stringify(expires),
        },
        type: {
          S: 'connect-session',
        },
        sess: {
          S: sess,
        },
        [this.options.ttlFieldName]: {
          N: this.getTTLValue().toString(),
        },
      },
    };
    await this.client.putItem(params).promise();
  }

  /**
   * Cleans up expired sessions
   *
   * @param {Function} fn
   * @api public
   */

  async reap() {
    const now = Math.floor(Date.now() / 1000);
    const options = {
      endkey: '[' + now + ',{}]',
    };
    const params = {
      TableName: this.table,
      ScanFilter: {
        expires: {
          AttributeValueList: [
            {
              N: now.toString(),
            },
          ],
          ComparisonOperator: 'LT',
        },
      },
      AttributesToGet: [this.hashKey],
    };
    try {
      const data = await this.client.scan(params).promise()
      for await (const item of data.Items || []) {
        await this.destroy(item[this.hashKey].S || '');
      }
    } catch (err) {
      console.log(err)
      throw err;
    }


  }

  /**
  * Destroy the session associated with the given `sid`.
  *
  * @param {String} sid
  * @param {Function} fn
  * @api public
  */
  async destroy(sid: string, fn?: (err?: any) => void) {

    await this.client.deleteItem({
      TableName: this.table, Key: {
        [this.hashKey]: {
          S: sid = this.prefix + sid
        }
      }
    }).promise();

    fn && fn(null)

  }

  /**
   * Touches the session row to update it's expire value.
   * @param  {String}   sid  Session id.
   * @param  {Object}   sess Session object.
   * @param  {Function} fn   Callback.
   */
  async touch(sid: string, sess: any, fn?: (err?: any, updated?: any) => void) {
    sid = this.prefix + sid;
    const expires = this.getExpiresValue(sess);
    const params = {
      TableName: this.table,
      Key: {
        [this.hashKey]: {
          S: sid,
        }
      },
      UpdateExpression: 'set expires = :e, #ttl = :ttl',
      ExpressionAttributeNames: {
        '#ttl': this.options.ttlFieldName
      },
      ExpressionAttributeValues: {
        ':e': {
          N: JSON.stringify(expires),
        },
        ':ttl': {
          N: this.getTTLValue().toString()
        }

      },
      ReturnValues: 'UPDATED_NEW',
    };

    try {
      const updated = await this.client.updateItem(params).promise();
      fn && fn(null, updated);
      return updated;
    } catch (error) {
      fn && fn(error);
      throw error;
    }
  }

  /**
  * Calculates the expire value based on the configuration.
  * @param  {Object} sess Session object.
  * @return {Integer}      The expire on timestamp.
  */
  getExpiresValue(sess: any) {
    const expires =
      typeof sess.cookie.maxAge === 'number'
        ? +new Date() + sess.cookie.maxAge
        : +new Date() + oneDayInMilliseconds;
    return Math.floor(expires / 1000);
  }

  /**
   * get value for set TTL 
   * @returns 
   */
  getTTLValue() {
    const now = new Date().getTime();
    return Math.floor(now / 1000) + this.options.ttl;
  }
  /**
  * Clear intervals
  *
  * @api public
  */
  clearInterval() {
    if (this._reap) clearInterval(this._reap);
  }
  // regenerate(re: any, fn: Function) {

  //   console.log(re)
  //   fn(re)
  // }

}


src/__test__/app.ts
import express from 'express';
import expressSession from 'express-session';
import connectSession from "../../src/index";
import AWS from 'aws-sdk';


export function app(maxAge = 1000) {
  const _app = express();
  const options = {
    saveUninitialized: true,
    resave: true,
    client: new AWS.DynamoDB({
      endpoint: new AWS.Endpoint('http://localhost:8000'),
      region: 'us-east-1'
    })
  };
  const DynamoDBStore = connectSession(expressSession)
  _app.use(expressSession({
    cookie: {
      maxAge,
    },
    store: new DynamoDBStore(options), secret: 'keyboard cat'
  }));

  _app.use(expressSession({
    secret: 'keyboard cat',
    resave: false,
    saveUninitialized: true,
    cookie: { secure: true }
  }))
  // _app.use((err: Error, req: express.Request, res: express.Response, next: express.NextFunction) => {
  //   res.json(err)
  // })
  _app.get('/', (req: express.Request, res: express.Response) => {
    (req.session as any).test = true;
    res.status(200).json((req as any).session);
  });

  _app.get('/inc', (req: express.Request, res: express.Response) => {
    (req.session as any).count = ((req.session as any).count || 0) + 1;
    res.status(200).json((req as any).session);
  });

  // _app.get('/inc', (req: express.Request, res: express.Response) => {
  //   (req.session as any).count = ((req.session as any).count || 0) + 1;
  //   res.status(200).json((req as any).session);
  // });


  _app.get('/reset', async (req: express.Request, res: express.Response) => {
    const oldId = req.session.id;
    try {
      req.session.regenerate(() => {
        (req.session as any).oldId = oldId
        res.json({
          ...req.session
        })
      });
    } catch (err) {
      console.log(err)
    }

  });


  return _app;
}
